### ISSUE TITLE
AttributeError: 'Artifact' object has no attribute 'group'

### ISSUE THREAD

**Issue by mhsjlw** [jiptool/jip#34]
When running most commands like `jip list` and `jip freeze > pom.xml` I get this error:

```
mhsjlw@arch ~/D/jpython-test> jip freeze > pom.xml
Traceback (most recent call last):
  File "/usr/bin/jip", line 9, in <module>
    load_entry_point('jip==0.9.9.dev1', 'console_scripts', 'jip')()
  File "/usr/lib/python3.5/site-packages/jip-0.9.9.dev1-py3.5.egg/jip/main.py", line 62, in main
    commands[cmd](**args)
  File "/usr/lib/python3.5/site-packages/jip-0.9.9.dev1-py3.5.egg/jip/commands.py", line 45, in wrapper
    index_manager.initialize()
  File "/usr/lib/python3.5/site-packages/jip-0.9.9.dev1-py3.5.egg/jip/index.py", line 89, in initialize
    artifacts = pickle.loads(pickledata)
  File "/usr/lib/python3.5/site-packages/jip-0.9.9.dev1-py3.5.egg/jip/maven.py", line 71, in __hash__
    return self.group.__hash__()*13+self.artifact.__hash__()*7+self.version.__hash__()
AttributeError: 'Artifact' object has no attribute 'group'
```

I've installed jip by cloning the repository and running `sudo setup.py install`
Any ideas on what's wrong?

**EDIT**: It seems that virtualenv also fails with a similar error:

```
mhsjlw@arch ~/D/jpython-test> virtualenv -p /usr/local/bin/jython jython-env
Running virtualenv with interpreter /usr/local/bin/jython
Traceback (most recent call last):
  File "/usr/lib/python3.5/site-packages/virtualenv.py", line 25, in <module>
    import distutils.sysconfig
  File "/home/mhsjlw/jython2.7.0/Lib/distutils/sysconfig.py", line 28, in <module>
    project_base = os.path.dirname(os.path.realpath(sys.executable))
  File "/home/mhsjlw/jython2.7.0/Lib/posixpath.py", line 367, in realpath
    path, ok = _joinrealpath('', filename, {})
  File "/home/mhsjlw/jython2.7.0/Lib/posixpath.py", line 373, in _joinrealpath
    if isabs(rest):
  File "/home/mhsjlw/jython2.7.0/Lib/posixpath.py", line 61, in isabs
    return s.startswith('/')
AttributeError: 'NoneType' object has no attribute 'startswith'

```

**Comment by sunng87** [comment]
Hi @mhsjlw , it seems two different issues. With jython you have an installation issue, while python3.5 it's a runtime issue.

I cannot produce the jython installation issue. I'm using the jython and virtualenv from arch community repo and the installation works for me.

**Comment by wikier** [comment]
I can't reproduce it either.

**Comment by xhkz** [comment]
Hi @sunng87, @wikier ,

Thank you for your great work.

I had same issue today.

(jyenv)  ~/code/jip >jip version
jip [Version] jip 0.9.6, jython 2.7.0 (default:9987c746f838, Apr 29 2015, 02:25:11) 
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)]
(jyenv)  ~/code/jip >jip install commons-io:commons-io:2.4
jip [Skipped] pom file not found at ~/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.pom
jip [Skipped] Pom file not found at http://archiva/repository/snapshots/commons-io/commons-io/2.4/commons-io-2.4.pom
jip [Skipped] pom file not found at ~/.m2/repository/org/apache/commons/commons-parent/25/commons-parent-25.pom
jip [Skipped] Pom file not found at http://archiva/repository/snapshots/org/apache/commons/commons-parent/25/commons-parent-25.pom
jip [Downloading] jar from http://archiva/repository/internal/commons-io/commons-io/2.4/commons-io-2.4.jar
jip [Finished] http://archiva/repository/internal/commons-io/commons-io/2.4/commons-io-2.4.jar downloaded 
jip [Finished] dependencies resolved
(jyenv)  ~/code/jip >jip version                          
Traceback (most recent call last):
  File "/home/xin/.venvs/jyenv/bin/jip", line 8, in <module>
    sys.exit(
  File "__pyclasspath__/jip/main.py", line 60, in main
  File "__pyclasspath__/jip/commands.py", line 47, in wrapper
  File "__pyclasspath__/jip/index$py.class", line 89, in initialize
  File "/home/xin/tools/jython/Lib/pickle.py", line 1382, in loads
    return Unpickler(file).load()
  File "/home/xin/tools/jython/Lib/pickle.py", line 858, in load
    dispatch[key](self)
  File "/home/xin/tools/jython/Lib/pickle.py", line 858, in load
    dispatch[key](self)
  File "/home/xin/tools/jython/Lib/pickle.py", line 1198, in load_setitem
    dict[key] = value
  File "__pyclasspath__/jip/maven.py", line 71, in __hash__
AttributeError: 'Artifact' object has no attribute 'group'


jip is working in a clean new virtualenv, however, after install any dependency with jip like what in the log above, and downloading the correct jar file, jip stops working.

**Comment by wikier** [comment]
Can't reproduce it:

```
wikier@wolfcastle:~$ virtualenv /tmp/ve
wikier@wolfcastle:~$ source /tmp/ve/bin/activate
(ve)wikier@wolfcastle:~$ pip install jip
Collecting jip
  Downloading jip-0.9.8.zip
Collecting requests (from jip)
  Downloading requests-2.11.0-py2.py3-none-any.whl (514kB)
    100% |████████████████████████████████| 516kB 1.0MB/s 
Building wheels for collected packages: jip
  Running setup.py bdist_wheel for jip
  Stored in directory: /home/wikier/.cache/pip/wheels/cd/94/5b/76fb90b1fe9312fe85fee452a17b4b53a0a9baae48c90762c2
Successfully built jip
Installing collected packages: requests, jip
Successfully installed jip-0.9.8 requests-2.11.0
You are using pip version 7.1.2, however version 8.1.2 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
(ve)wikier@wolfcastle:~$ **jip version**
jip [Version] jip 0.9.8, jython 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609]
(ve)wikier@wolfcastle:~$ **jip install commons-io:commons-io:2.4**
jip [Checking] pom file /home/wikier/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.pom
jip [Checking] pom file /home/wikier/.m2/repository/org/apache/commons/commons-parent/25/commons-parent-25.pom
jip [Checking] pom file /home/wikier/.m2/repository/org/apache/apache/9/apache-9.pom
jip [Checking] jar package from local
jip [Downloading] /home/wikier/.m2/repository/commons-io/commons-io/2.4/commons-io-2.4.jar
jip [Finished] /home/wikier/.jip/cache/commons-io/commons-io/commons-io-2.4.jar completed
jip [Finished] dependencies resolved
(ve)wikier@wolfcastle:~$ **jip version**  
jip [Version] jip 0.9.8, jython 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609]
```

**Comment by xhkz** [comment]
Hi @wikier ,

Thanks for your reply.

May I know where can I get your jython 2.7.12?
I am using jython 2.7 from http://www.jython.org/downloads.html. While the jython's difference might not be the reason for this issue, I could give it a test with my environment using the same jython.

```
(jyenv)  ~/code/jip >jip version
jip [Version] jip 0.9.6, jython 2.7.0 (default:9987c746f838, Apr 29 2015, 02:25:11) 
[Java HotSpot(TM) 64-Bit Server VM (Oracle Corporation)]

(ve)wikier@wolfcastle:~$ jip version 
jip [Version] jip 0.9.8, jython 2.7.12 (default, Jul  1 2016, 15:12:24) 
[GCC 5.4.0 20160609]
```

**Comment by shayneoAtNorwood** [comment]
I'm getting the same error as well. I wouldnt have a clue where to start with it :(

**Comment by testlnord** [comment]
Got the same error. 

I'm trying to run [JayDeBeAPI](https://github.com/baztian/jaydebeapi) tests with Python 2.7. After successful installation of 2 packages jip failed to install ``org.xerial:sqlite-jdbc:3.7.2``. 

Now it can't even say its version:
``` 
(jaydebeapi-py27)arkady@arkady-notebook:~/Documents/projects/jaydebeapi$ jip version
Traceback (most recent call last):
  File "/home/arkady/.virtualenvs/jaydebeapi-py27/bin/jip", line 9, in <module>
    load_entry_point('jip==0.9.3', 'console_scripts', 'jip')()
  File "/home/arkady/.virtualenvs/jaydebeapi-py27/local/lib/python2.7/site-packages/jip/main.py", line 62, in main
    commands[cmd](**args)
  File "/home/arkady/.virtualenvs/jaydebeapi-py27/local/lib/python2.7/site-packages/jip/commands.py", line 45, in wrapper
    index_manager.initialize()
  File "/home/arkady/.virtualenvs/jaydebeapi-py27/local/lib/python2.7/site-packages/jip/index.py", line 89, in initialize
    artifacts = pickle.loads(pickledata)
  File "/usr/lib/python2.7/pickle.py", line 1382, in loads
    return Unpickler(file).load()
  File "/usr/lib/python2.7/pickle.py", line 858, in load
    dispatch[key](self)
  File "/usr/lib/python2.7/pickle.py", line 1198, in load_setitem
    dict[key] = value
  File "/home/arkady/.virtualenvs/jaydebeapi-py27/local/lib/python2.7/site-packages/jip/maven.py", line 71, in __hash__
    return self.group.__hash__()*13+self.artifact.__hash__()*7+self.version.__hash__()
AttributeError: 'Artifact' object has no attribute 'group'
(jaydebeapi-py27)arkady@arkady-notebook:~/Documents/projects/jaydebeapi$ jip install org.xerial:sqlite-jdbc:3.7.2
```

**Comment by testlnord** [comment]
It seems like ``.jip_index`` was corrupted somehow. When I delete it everything works again, until I install another package.

**Comment by sunng87** [comment]
Confirmed it's caused by custom __hash__ definition for jip internal data type. And Python pickle module updated to call __hash__ before some property loaded. Will prepare a fix soon.

**Comment by sunng87** [comment]
Released in 0.9.9. Please update.
